pkg/reconciler/instances/istio/actions/actions.go

```diff
@@ -0,0 +1,146 @@
+}
+
+type webhookPatchJSONValue struct {
+	Key      string   `json:"key"`
+	Operator string   `json:"operator"`
+	Values   []string `json:"values"`
+}
+
+//go:generate mockery -name=IstioPerformer
+// IstioPerformer performs actions on Istio component on the cluster.
+type IstioPerformer interface {
+
+	// Install Istio on the cluster.
+	Install() error
+
+	// PatchMutatingWebhook configuration.
+	PatchMutatingWebhook() error
+}
+
+// DefaultIstioPerformer provides a default implementation of IstioPerformer.
+type DefaultIstioPerformer struct {
+	commander     istioctl.Commander
+	istioctlPath  string
+	kubeConfig    string
+	istioOperator string
+	kubeClient    kubernetes.Client
+	logger        *zap.SugaredLogger
+}
+
+// NewDefaultIstioPerformer creates a new instance of the DefaultIstioPerformer.
+func NewDefaultIstioPerformer(kubeConfig, manifest string, kubeClient kubernetes.Client, logger *zap.SugaredLogger, cmder istioctl.Commander) (*DefaultIstioPerformer, error) {
+	istioctlPath := resolveIstioctlPath()
+	if istioctlPath == "" {
+		return nil, errors.New("Istioctl binary could not be found under ISTIOCTL_PATH env variable")
+	}
+
+	istioOperator := extractIstioOperatorContextFrom(manifest)
+	if istioOperator == "" {
+		return nil, errors.New("Istio Operator definition could not be found in manifest")
+	}
+
+	return &DefaultIstioPerformer{
+		istioctlPath:  istioctlPath,
+		kubeConfig:    kubeConfig,
+		istioOperator: istioOperator,
+		kubeClient:    kubeClient,
+		logger:        logger,
+		commander:     cmder,
+	}, nil
+}
+
+func (c *DefaultIstioPerformer) Install() error {
+	istioOperatorPath, istioOperatorCf, err := file.CreateTempFileWith(c.istioOperator)
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		cleanupErr := istioOperatorCf()
+		if cleanupErr != nil {
+			c.logger.Error(cleanupErr)
+		}
+	}()
+
+	kubeconfigPath, kubeconfigCf, err := file.CreateTempFileWith(c.kubeConfig)
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		cleanupErr := kubeconfigCf()
+		if cleanupErr != nil {
+			c.logger.Error(cleanupErr)
+		}
+	}()
+
+	err = c.commander.Install(c.istioctlPath, istioOperatorPath, kubeconfigPath)
+	if err != nil {
+		return errors.Wrap(err, "Error occurred when calling istioctl")
+	}
+
+	return nil
+}
+
+func (c *DefaultIstioPerformer) PatchMutatingWebhook() error {
+	patchContent := []webhookPatchJSON{{
+		Op:   "add",
+		Path: "/webhooks/4/namespaceSelector/matchExpressions/-",
+		Value: webhookPatchJSONValue{
+			Key:      "gardener.cloud/purpose",
+			Operator: "NotIn",
+			Values: []string{
+				"kube-system",
+			},
+		},
+	}}
+
+	patchContentJSON, err := json.Marshal(patchContent)
+	if err != nil {
+		return err
+	}
+
+	err = c.kubeClient.PatchUsingStrategy("MutatingWebhookConfiguration", "istio-sidecar-injector", "istio-system", patchContentJSON, types.JSONPatchType)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func resolveIstioctlPath() string {
+	return os.Getenv(istioctlBinaryPathEnvKey)
```

Comment:
For local testig, it would be good to have an option to use also a locally installed istioctl command as fallback, like

    func resolveIstioctlPath (string, err) {
       istioctlPath, ok := os.Lookup(istioctlBinaryPathEnvKey)
       if ok {
         if file.Exists(istioctlPath) {
           return istioctlPath, nil
         }
         return "", fmt. Errorf("istioctl binary '%s' defined in env-var '%s' but file does not exist", istioctlPath, istioctlBinaryPathEnvKey)
       }
       istioctlPath, err := exec.Command("which istioctl").Output()
       if err != nil {
           return "", fmt.Errorf("istiotctl binary not defined in env-var '%' and not available in $PATH", istioctlBinaryPathEnvKey)
       }
       return istioctlPath, nil
    }

Reply:
It would be cool to have such quality-of-life improvement, BUT:
- Windows has it's own equivalent to `which` command
- the binary does not have to be called exactly `istioctl` - it may contain prefixes or suffixes
- you may have multiple `istioctl` binaries available on your machine - which one should be chosen?
We might come back to the discussion how to handle these, but for now I think it's not crucial. As a user you simply have to set your env var on the local machine and this Reconciler will still work!